drivers/pwm/pwm_rpi_pico_slice.c

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright (c) 2022, Joep Buruma
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#define DT_DRV_COMPAT raspberrypi_pico_pwm_slice
+
+#include <device.h>
+#include <drivers/pwm.h>
+#include <drivers/pinctrl.h>
+#include <logging/log.h>
+LOG_MODULE_REGISTER(pwm_rpi_pico_slice, CONFIG_PWM_LOG_LEVEL);
+
+/* pico-sdk includes */
+#include <hardware/pwm.h>
+#include <hardware/structs/pwm.h>
+
+#define PWM_RPI_PICO_COUNTER_TOP_MAX UINT16_MAX
+
+struct pwm_rpi_divider_setting {
+	uint8_t integral;
+	uint8_t frac;
+};
+
+struct pwm_rpi_config {
+	/* pwm_controller is the start address of the pwm peripheral.
+	 * This address is used to get a refference to pwm_slice_hw,
+	 * which contains all registers for configuring a single slice
+	 */
+	pwm_hw_t *pwm_controller;
+	uint8_t pwm_slice_nr;
+	struct pwm_rpi_divider_setting divider;
+	const struct pinctrl_dev_config *pcfg;
+};
+
+struct pwm_rpi_data {
+	struct pwm_slice_hw *slice;
+};
+
+static float pwm_rpi_get_clkdiv(const struct device *dev)
+{
+	const struct pwm_rpi_config *cfg = dev->config;
+
+	/* the divider is a fixed point 8.4 convert to float for use in pico-sdk */
+	return (float)cfg->divider.integral + (float)cfg->divider.integral / 16.0;
+}
+
+static int pwm_rpi_get_cycles_per_sec(const struct device *dev, uint32_t ch, uint64_t *cycles)
+{
+	float f_clock_in = (float)sys_clock_hw_cycles_per_sec();
+
+	/* No need to check for divide by 0 since the minimum value of
+	 * returned is 1
+	 */
+	*cycles = (uint64_t)(f_clock_in / pwm_rpi_get_clkdiv(dev));
+	return 0;
+}
+
+/* The pico_sdk only allows setting the polarity of both channels at once.
+ * This is a convenience function to make setting the polarity of a single
+ * channel easier.
+ */
+static void pwm_rpi_set_channel_polarity(const struct device *dev, uint32_t chan, bool inverted)
+{
+	struct pwm_rpi_data *data = dev->data;
+	const struct pwm_rpi_config *cfg = dev->config;
+	bool pwm_polarity_a = (data->slice->csr & PWM_CH0_CSR_A_INV_BITS) > 0;
```

Comment:
The bool to bit conversion (using >) isn't necessary, as `pwm_set_output_polarity` converts it using `bool_to_bit`

Reply:
This is not a conversion from bool to but, but from bit to bool. This is then used to later invert the selected channel.
Then later on `pwm_set_output_polarity` calls `bool_to_bit` to convert it to a bit again.